### PR TITLE
Made num_partitions optional

### PR DESCRIPTION
**Pull request by mahajanadhitya**

### CODE REVIEW

File: src/confluent_kafka/src/AdminTypes.c [hunk 1]
```diff
@@ -73,19 +73,29 @@ static int NewTopic_init (PyObject *self0, PyObject *args,
...
-        if (!PyArg_ParseTupleAndKeywords(args, kwargs, "si|iOO", kws,
+        if (!PyArg_ParseTupleAndKeywords(args, kwargs, "s|iiOO", kws,
                                          &topic, &self->num_partitions,
                                          &self->replication_factor,
                                          &self->replica_assignment,
                                          &self->config))
                 return -1;
 
 
+        if(self->replica_assignment){
```
Comment (pranavrth) [inline src/confluent_kafka/src/AdminTypes.c:90]: Don't raise this error here. It should be in librdkafka or from the broker.

File: src/confluent_kafka/src/Admin.c [hunk 2]
```diff
@@ -510,6 +511,19 @@ static PyObject *Admin_create_topics (Handle *self, PyObject *args,
                                 goto err;
                         }
                 }
+                if (!newt->replica_assignment) {
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:514]: Don't raise this error here. It should be in librdkafka or from the broker.

File: src/confluent_kafka/src/Admin.c [hunk 3]
```diff
@@ -495,7 +495,8 @@ static PyObject *Admin_create_topics (Handle *self, PyObject *args,
                         if (!Admin_set_replica_assignment(
                                     "CreateTopics", (void *)c_objs[i],
                                     newt->replica_assignment,
-                                    newt->num_partitions, newt->num_partitions,
+                                    PyList_Size(newt->replica_assignment),
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:498]: Extract to a variable.

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,7 +5,7 @@
 - Added `set_sasl_credentials`. This new method (on the Producer, Consumer, and AdminClient) allows modifying the stored
   SASL PLAIN/SCRAM credentials that will be used for subsequent (new) connections to a broker (#1511).
 - Wheels for Linux / arm64 (#1496).
-
+- Made support for use of Default num_partitions and replication_factor in NewTopic according to KIP - 464
```
Comment (pranavrth) [inline CHANGELOG.md:8]: Added support for Default num_partitions in CreateTopics Admin API according to KIP-464.

**Review by pranavrth — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 4 above)
Comment (pranavrth) [reply CHANGELOG.md:8]: This field is optional now.

File: src/confluent_kafka/src/AdminTypes.c [hunk 5]
```diff
@@ -73,19 +73,33 @@ static int NewTopic_init (PyObject *self0, PyObject *args,
...
```
Comment (pranavrth) [inline src/confluent_kafka/src/AdminTypes.c:90]: Let's not add these check. Keep the existing checks only. These checks should be added in librdkafka or broker side as these are applicable only for v4 of CreateAPI but we are supporting all the versions.

File: src/confluent_kafka/src/AdminTypes.c [hunk 6]
```diff
@@ -202,7 +223,12 @@ NewTopic_richcompare (NewTopic *self, PyObject *o2, int op) {
 
 static long NewTopic_hash (NewTopic *self) {
         PyObject *topic = cfl_PyUnistr(_FromString(self->topic));
-        long r = PyObject_Hash(topic) ^ self->num_partitions;
+        long r;
+        if( self->num_partitions==-1 ){
```
Comment (pranavrth) [inline src/confluent_kafka/src/AdminTypes.c:227]: use consistent style

File: src/confluent_kafka/src/AdminTypes.c [hunk 7]
```diff
@@ -147,9 +162,15 @@ static PyMemberDef NewTopic_members[] = {
 
 
 static PyObject *NewTopic_str0 (NewTopic *self) {
-        return cfl_PyUnistr(
+        if(self->num_partitions != -1){
```
Comment (pranavrth) [inline src/confluent_kafka/src/AdminTypes.c:165]: Better to have check for exception instead of normal case.

Also, style. Space between ) and {

**Review by pranavrth — CHANGES_REQUESTED** [review]

File: src/confluent_kafka/src/Admin.c [hunk 8]
```diff
@@ -483,19 +483,11 @@ static PyObject *Admin_create_topics (Handle *self, PyObject *args,
...
-                                                "mutually exclusive");
-                                i++;
-                                goto err;
-                        }
-
                         if (!Admin_set_replica_assignment(
                                     "CreateTopics", (void *)c_objs[i],
                                     newt->replica_assignment,
-                                    newt->num_partitions, newt->num_partitions,
+                                    PyList_Size(newt->replica_assignment),
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:489]: Dont change existing logic. make backward compatible.

File: CHANGELOG.md [hunk 9]
```diff
@@ -5,7 +5,7 @@
 - Added `set_sasl_credentials`. This new method (on the Producer, Consumer, and AdminClient) allows modifying the stored
   SASL PLAIN/SCRAM credentials that will be used for subsequent (new) connections to a broker (#1511).
 - Wheels for Linux / arm64 (#1496).
-
+- Added support for Default num_partitions(the field is optional now) in CreateTopics Admin API according to KIP-464.
```
Comment (pranavrth) [inline CHANGELOG.md:8]: Remove KIP-464 part

File: src/confluent_kafka/src/Admin.c [hunk 10]
```diff
@@ -483,19 +483,11 @@ static PyObject *Admin_create_topics (Handle *self, PyObject *args,
                 }
 
                 if (newt->replica_assignment) {
-                        if (newt->replication_factor != -1) {
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:486]: Keep it as it is.

**Review by pranavrth — CHANGES_REQUESTED** [review]

File: src/confluent_kafka/src/AdminTypes.c [hunk 11]
```diff
@@ -125,7 +139,8 @@ static PyMemberDef NewTopic_members[] = {
         { "topic", T_STRING, offsetof(NewTopic, topic), READONLY,
           ":py:attribute:topic - Topic name (string)" },
         { "num_partitions", T_INT, offsetof(NewTopic, num_partitions), 0,
-          ":py:attribute: Number of partitions (int)" },
+          ":py:attribute: [OPTIONAL] Number of partitions (int).\n"
+          "Must be set to -1 or not be provided if a replica_assignment is specified" },
```
Comment (pranavrth) [inline src/confluent_kafka/src/AdminTypes.c:143]: Check this - https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/NewTopic.java

File: src/confluent_kafka/src/Admin.c [hunk 12]
```diff
@@ -492,10 +492,17 @@ static PyObject *Admin_create_topics (Handle *self, PyObject *args,
                                 goto err;
                         }
 
+                        int t;
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:495]: proper name.

File: src/confluent_kafka/src/AdminTypes.c [hunk 13]
```diff
@@ -125,7 +127,8 @@ static PyMemberDef NewTopic_members[] = {
         { "topic", T_STRING, offsetof(NewTopic, topic), READONLY,
           ":py:attribute:topic - Topic name (string)" },
         { "num_partitions", T_INT, offsetof(NewTopic, num_partitions), 0,
-          ":py:attribute: Number of partitions (int)" },
+          ":py:attribute: Number of partitions (int).\n"
+          "Must be set to -1 if a replica_assignment is specified" },
```
Comment (pranavrth) [inline src/confluent_kafka/src/AdminTypes.c:131]: If we are enforcing this, then we don't need this condition.

```
if (newt->num_partitions == -1) {
        t = PyList_Size(newt->replica_assignment);
} else {
        t = newt->num_partitions;
}
```

We can use what's written in Java. In this way, its a suggestion.

> The number of partitions for the new topic or -1 if a replica assignment has been specified.

File: CHANGELOG.md [hunk 14]
```diff
@@ -5,7 +5,7 @@
 - Added `set_sasl_credentials`. This new method (on the Producer, Consumer, and AdminClient) allows modifying the stored
   SASL PLAIN/SCRAM credentials that will be used for subsequent (new) connections to a broker (#1511).
 - Wheels for Linux / arm64 (#1496).
-
+- Added support for Default num_partitions(the field is optional now) in CreateTopics Admin API.
```
Comment (pranavrth) [inline CHANGELOG.md:8]: Remove "(the field is optional now)"

**Review by pranavrth — CHANGES_REQUESTED** [review]

File: src/confluent_kafka/src/AdminTypes.c (hunk 13 above)
Comment (mahajanadhitya) [reply src/confluent_kafka/src/AdminTypes.c:131]: We are not enforcing to keep backward compatibility, we need the check so prior code does not break.

File: src/confluent_kafka/src/Admin.c (hunk 12 above)
Comment (mahajanadhitya) [reply src/confluent_kafka/src/Admin.c:495]: temp_partitions ?

**Review by pranavrth — APPROVED** [review]
Good Job.

LGTM!

**Pull request merged**